model/src/lib.rs

```diff
@@ -61,44 +55,31 @@ impl Signature {
         }
     }
 
-    fn signing_digest_typed_data(
+    pub fn validate(
+        &self,
         domain_separator: &DomainSeparator,
-        digest: &[u8; 32],
-    ) -> [u8; 32] {
-        let mut hash_data = [0u8; 66];
-        hash_data[0..2].copy_from_slice(&[0x19, 0x01]);
-        hash_data[2..34].copy_from_slice(&domain_separator.0);
-        hash_data[34..66].copy_from_slice(digest);
-        signing::keccak256(&hash_data)
+        struct_hash: &[u8; 32],
+    ) -> Option<H160> {
+        let message = hashed_eip712_message(domain_separator, struct_hash);
+        let recovery = Recovery::new(message, self.v as u64, self.r, self.s);
+        let (signature, recovery_id) = recovery.as_signature()?;
+        signing::recover(&message, &signature, recovery_id).ok()
     }
 
-    pub fn signing_digest_message(
+    pub fn sign(
         domain_separator: &DomainSeparator,
-        digest: &[u8; 32],
-    ) -> [u8; 32] {
-        let mut hash_data = [0u8; 92];
-        hash_data[0..28].copy_from_slice(b"\x19Ethereum Signed Message:\n64");
-        hash_data[28..60].copy_from_slice(&domain_separator.0);
-        hash_data[60..92].copy_from_slice(digest);
-        signing::keccak256(&hash_data)
-    }
-
-    fn signing_digest(&self, domain_separator: &DomainSeparator, digest: &[u8; 32]) -> [u8; 32] {
-        // This is fallback for wallets that don't support EIP712
-        if self.v & 0x80 == 0 {
-            Signature::signing_digest_typed_data(domain_separator, digest)
-        } else {
-            Signature::signing_digest_message(domain_separator, digest)
```

Comment:
Have we somehow generalized the signing scheme to no longer need both scenarios here?

Reply:
No. This PR assumes eip712 and I am going to make another one to add EthSign support back.